File: implementations/go/ast.go
package rcl

type Node interface {
	Kind() string
}

type DocumentNode struct {
	NodeKind string      `json:"kind"`
	Blocks   []BlockNode `json:"blocks"`
	RootValue Node       `json:"root_value,omitempty"`
}

func (n DocumentNode) Kind() string { return n.NodeKind }

type BlockNode struct {
	NodeKind    string              `json:"kind"`
	Name        string              `json:"name"`
	Argument    *string             `json:"argument,omitempty"`
	Properties  map[string]Node     `json:"properties"`
	Blocks      map[string]BlockNode `json:"blocks"`
	NamedBlocks []BlockNode         `json:"named_blocks,omitempty"`
}

func (n BlockNode) Kind() string { return n.NodeKind }

type StringNode struct {
	NodeKind string `json:"kind"`
	Value    string `json:"value"`
}

func (n StringNode) Kind() string { return n.NodeKind }

type NumberNode struct {
	NodeKind string  `json:"kind"`
	Value    float64 `json:"value"`
}

func (n NumberNode) Kind() string { return n.NodeKind }

type BooleanNode struct {
	NodeKind string `json:"kind"`
	Value    bool   `json:"value"`
}

func (n BooleanNode) Kind() string { return n.NodeKind }

type ArrayNode struct {
	NodeKind string `json:"kind"`
	Elements []Node `json:"elements"`
}

func (n ArrayNode) Kind() string { return n.NodeKind }

File: implementations/csharp/src/Ast.cs
using System.Collections.Generic;

namespace RCLImpl {
public interface IAstNode { string Kind { get; } }

public sealed class StringNode : IAstNode { public string Kind { get { return "string"; } } public string Value; public StringNode(string v){ Value=v; } }
public sealed class NumberNode : IAstNode { public string Kind { get { return "number"; } } public double Value; public NumberNode(double v){ Value=v; } }
public sealed class BooleanNode : IAstNode { public string Kind { get { return "boolean"; } } public bool Value; public BooleanNode(bool v){ Value=v; } }
public sealed class ArrayNode : IAstNode { public string Kind { get { return "array"; } } public List<IAstNode> Elements = new List<IAstNode>(); }

public sealed class BlockNode : IAstNode {
  public string Kind { get { return "block"; } }
  public string Name;
  public string Argument;
  public Dictionary<string, IAstNode> Properties = new Dictionary<string, IAstNode>();
  public List<string> PropertyOrder = new List<string>();
  public List<BlockNode> Blocks = new List<BlockNode>();
  public List<BlockNode> NamedBlocks = new List<BlockNode>();
  public BlockNode(string n, string a){ Name=n; Argument=a; }
}

public sealed class DocumentNode : IAstNode {
  public string Kind { get { return "document"; } }
  public List<BlockNode> Blocks = new List<BlockNode>();
  public IAstNode RootValue;
}
}
